### PR TITLE
Support for specifying custom date format for date and timestamp types.

### PR DESCRIPTION
**Pull request by HyukjinKwon**
https://github.com/databricks/spark-csv/issues/279
https://github.com/databricks/spark-csv/issues/262
https://github.com/databricks/spark-csv/issues/266

This PR adds the support to specify custom date format for `DateType` and `TimestampType`.

For `TimestampType`, this uses the given format to infer schema and also to convert the values
For `DateType`, this uses the given format to convert the values.
If the `dateFormat` is not given, then it works with `Timestamp.valueOf()` and `Date.valueOf()` for backwords compatibility.
When it's given, then it uses `SimpleDateFormat` for parsing data.

In addition, `IntegerType`, `DoubleType` and `LongType` have a higher priority than `TimestampType` in type inference. This means even if the given format is `yyyy` or `yyyy.MM`, it will be inferred as `IntegerType` or `DoubleType`. Since it is type inference, I think it is okay to give such precedences.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
@falaki Just to let you know, the original functions, `tryParse...()`  were not mostly modified but just moved into `inferField()` as inner functions at `src/main/scala/com/databricks/spark/csv/util/InferSchema.scala`.

File: src/main/scala/com/databricks/spark/csv/DefaultSource.scala [hunk 1]
```diff
@@ -128,6 +128,8 @@ class DefaultSource
     val charset = parameters.getOrElse("charset", TextFile.DEFAULT_CHARSET.name())
     // TODO validate charset?
 
+    val dataFormat = parameters.getOrElse("charset", TextFile.DEFAULT_CHARSET.name())
```
Comment (falaki) [inline src/main/scala/com/databricks/spark/csv/DefaultSource.scala]: This line needs to be removed.

File: src/main/scala/com/databricks/spark/csv/util/InferSchema.scala [hunk 2]
```diff
@@ -76,7 +77,61 @@ private[csv] object InferSchema {
    */
   private[csv] def inferField(typeSoFar: DataType,
     field: String,
-    nullValue: String = ""): DataType = {
+    nullValue: String = "",
+    dateFormatter: SimpleDateFormat = null): DataType = {
+    def tryParseInteger(field: String): DataType = if ((allCatch opt field.toInt).isDefined) {
```
Comment (falaki) [inline src/main/scala/com/databricks/spark/csv/util/InferSchema.scala]: Indent is off for this entire block

**Comment by falaki** [comment]
@HyukjinKwon left one more comment. Otherwise looks good. I can merge this before I cut the branch tonight.

File: src/main/scala/com/databricks/spark/csv/util/InferSchema.scala (hunk 2 above)
Comment (HyukjinKwon) [inline src/main/scala/com/databricks/spark/csv/util/InferSchema.scala]: Um.. Do you mean the indentation correction as below?
```scala
  private[csv] def inferField(typeSoFar: DataType,
    field: String,
    nullValue: String = "",
    dateFormatter: SimpleDateFormat = null): DataType = {
    def tryParseInteger(field: String): DataType = if ((allCatch opt field.toInt).isDefined) {
      IntegerType
    } else {
      tryParseLong(field)
    }
...
```

```scala
  private[csv] def inferField(typeSoFar: DataType,
    field: String,
    nullValue: String = "",
    dateFormatter: SimpleDateFormat = null): DataType = {
      def tryParseInteger(field: String): DataType = if ((allCatch opt field.toInt).isDefined) {
        IntegerType
      } else {
        tryParseLong(field)
      }
...
```

Comment (falaki) [inline src/main/scala/com/databricks/spark/csv/util/InferSchema.scala]: Oh I see. The problem is for lines above:
```scala
  private[csv] def inferField(typeSoFar: DataType,
      field: String,
      nullValue: String = "",
      dateFormatter: SimpleDateFormat = null): DataType = {
    def tryParseInteger(field: String): DataType = if ((allCatch opt field.toInt).isDefined) {
      IntegerType
    } else {
      tryParseLong(field)
    }
```

Comment (HyukjinKwon) [inline src/main/scala/com/databricks/spark/csv/util/InferSchema.scala]: I see. Thanks!

**Comment by HyukjinKwon** [comment]
@falaki (although this is not the right thread to say this), what do you think about `uncompressed` and `none` options for compression codecs to set explicitly no compression?

This was merged in Spark, https://github.com/apache/spark/pull/11464 and I was working on this for this library as well. 

However, I just realised that this might not be some kind of what we must do identically for this, like `compression` option is supported for `codec` at Spark CSV data source but `compression` is not acceptable at this library.

**Comment by falaki** [comment]
Let's open an issue for it.

**Comment by HyukjinKwon** [comment]
Filed in https://github.com/databricks/spark-csv/issues/281.

**Comment by barrybecker4** [comment]
Thank you for adding this! I will pull and build a local snapshot until it is 1.4.0 officially releases.

**Comment by HyukjinKwon** [comment]
@barrybecker4 Maybe would you create a PR for that typo (also for other typos if you know)?

**Comment by 5ean** [comment]
I still face error in python. Maybe i did not use it correctly. Please kindly advise.
`df = sqlContext.read.format('com.databricks.spark.csv').options(header='true',dateFormat='dd/MM/yyyy hh:mm:ss').load('test.csv',schema = schema)`

schema has timestamp type.

And the string in csv file '"25/02/2014 00:00:00"

exception:

Caused by: java.lang.IllegalArgumentException: Timestamp format must be yyyy-mm-dd hh:mm:ss[.fffffffff]

**Comment by raam86** [comment]
The docs are not clear on how to do this so hopefully this can help:

https://stackoverflow.com/questions/43259485/how-to-load-csvs-with-timestamps-in-custom-format

```
spark.read
      .option("header", true)
      .option("inferSchema", true)
      .option("timestampFormat", "MM/dd/yyyy h:mm:ss a")
      .csv("PATH_TO_CSV")
```

**Comment by shatestest** [comment]
this is not working as expected.  https://stackoverflow.com/questions/55965978/how-to-set-jdbc-partitioncolumn-type-to-date-in-spark-2-4-1/55966481#55966481

**Pull request closed without merging**